internal/csi-addons/rbd/volumegroup.go

```diff
@@ -215,3 +216,150 @@ func (vs *VolumeGroupServer) DeleteVolumeGroup(
+// - resolve the existing volume group
+// - get the CSI-IDs of all volumes
+// - create a list of volumes that should be removed
+// - create a list of volume IDs that should be added
+// - remove the volumes from the group
+// - add the volumes to the group
+//
+// Also, MODIFY_VOLUME_GROUP_MEMBERSHIP does not exist, it is called
+// MODIFY_VOLUME_GROUP instead.
+func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
+	ctx context.Context,
+	req *volumegroup.ModifyVolumeGroupMembershipRequest,
+) (*volumegroup.ModifyVolumeGroupMembershipResponse, error) {
+	mgr := rbd.NewManager(vs.csiID, nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	// resolve the volume group
+	vg, err := mgr.GetVolumeGroupByID(ctx, req.GetVolumeGroupId())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"could not find volume group %q: %s",
+			req.GetVolumeGroupId(),
+			err.Error())
+	}
+	defer vg.Destroy(ctx)
+
+	beforeVolumes, err := vg.ListVolumes(ctx)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to list volumes of volume group %q: %v",
+			vg,
+			err)
+	}
+
+	// beforeIDs contains the csiID as key, volume as value
+	beforeIDs := make(map[string]types.Volume, len(beforeVolumes))
+	for _, vol := range beforeVolumes {
+		id, idErr := vol.GetID(ctx)
+		if idErr != nil {
+			return nil, status.Errorf(
+				codes.InvalidArgument,
+				"failed to get the CSI ID of volume %q: %v",
+				vol,
+				err)
+		}
+
+		beforeIDs[id] = vol
+	}
+
+	// check which volumes should not be part of the group
+	afterIDs := req.GetVolumeIds()
+	toRemove := make([]string, 0)
+	for id := range beforeIDs {
+		if !slices.Contains(afterIDs, id) {
+			toRemove = append(toRemove, id)
+		}
+	}
+
+	// check which volumes are new to the group
+	toAdd := make([]string, 0)
+	for _, id := range afterIDs {
+		if _, ok := beforeIDs[id]; !ok {
+			toAdd = append(toAdd, id)
+		}
+	}
+
+	// remove the volume that should not be part of the group
+	for _, id := range toRemove {
+		vol := beforeIDs[id]
+		err = vg.RemoveVolume(ctx, vol)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"failed to remove volume %q from volume group %q: %v",
+				vol,
+				vg,
+				err)
+		}
+	}
+
+	// add the new volumes to the group
+	for _, id := range toAdd {
+		vol, getErr := mgr.GetVolumeByID(ctx, id)
+		if getErr != nil {
+			return nil, status.Errorf(
+				codes.NotFound,
+				"failed to find a volume with CSI ID %q: %v",
+				id,
+				err)
+		}
+
+		err = vg.AddVolume(ctx, vol)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"failed to add volume %q to volume group %q: %v",
+				vol,
+				vg,
+				err)
+		}
+	}
+
+	csiVG, err := vg.ToCSI(ctx)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to convert volume group %q to CSI format: %v",
+			vg,
+			err)
+	}
+
+	return &volumegroup.ModifyVolumeGroupMembershipResponse{
+		VolumeGroup: csiVG,
```

Comment:
As discussed in the other (csi-addons [PR](https://github.com/csi-addons/kubernetes-csi-addons/pull/610#discussion_r1683963828), do we need to send a response with the volumegroup or just send an empty response similar to Delete operation>?

Reply:
It is like this in the specification of the CSI-Addons protocol for the operation, see https://github.com/csi-addons/spec/blob/main/volumegroup/README.md#modifyvolumegroupmembership.

The protocol that kubernetes-csi-addons uses to communicate between its Controller and sidecars can be different from the actual CSI-Addons Spec.